test/master-playlist-controller.test.js

```diff
@@ -1067,6 +1067,190 @@ function(assert) {
+  // timeout means we are on the last rendition)
+  segmentLoader.xhrOptions_.timeout = 60000;
+
+  // we need to wait 1 second from first byte receieved in order to consider aborting
+  this.requests[0].downloadProgress({
+    target: this.requests[0],
+    total: 100,
+    loaded: 1
+  });
+  this.clock.tick(1000);
+
+  // should abort request early because we don't have enough bandwidth
+  this.requests[0].downloadProgress({
+    target: this.requests[0],
+    total: 100,
+    // 1 bit per second
+    loaded: 2
+  });
+
+  this.clock.tick(1);
+
+  assert.equal(mediaChanges.length, 2, 'changed media');
+  assert.equal(mediaChanges[1].uri, 'media.m3u8', 'changed to media');
+  assert.equal(segmentRequest.uri.substring(segmentRequest.uri.length - 4),
+               '0.ts',
+               'requested first segment');
+  assert.equal(bandwidthupdateEvents, 2, 'two bandwidth update events');
+
+  // remove aborted request
+  this.requests.shift();
+
+  // 1ms for the cached segment response
+  this.clock.tick(1);
+
+  // 10 bytes in 1ms = 80 kbps
+  this.requests[0].response = new Uint8Array(10).buffer;
+  this.requests.shift().respond(200, null, '');
+
+  segmentLoader.mediaSource_.sourceBuffers[0].trigger('updateend');
+  this.clock.tick(1);
+
+  assert.equal(segmentLoader.bandwidth,
+               10 * Config.BANDWIDTH_VARIANCE + 1,
+               'bandwidth remains at aborted bandwidth');
+  assert.equal(mediaChanges.length, 2, 'did not change media');
+  assert.equal(bandwidthupdateEvents, 2, 'still two bandwidth update events');
+
+  segmentRequest = this.requests[0];
+  assert.equal(segmentRequest.uri.substring(segmentRequest.uri.length - 4),
+               '1.ts',
+               'requested second segment');
+
+  // must have two segment requests before we consider upswitching, so mimic another cache
+
+  // 1ms for the cached segment response
+  this.clock.tick(1);
+
+  // 10 bytes in 1ms = 80 kbps
+  this.requests[0].response = new Uint8Array(10).buffer;
+  this.requests.shift().respond(200, null, '');
+
+  segmentLoader.mediaSource_.sourceBuffers[0].trigger('updateend');
+  this.clock.tick(1);
+
+  assert.equal(segmentLoader.bandwidth,
+               10 * Config.BANDWIDTH_VARIANCE + 1,
+               'bandwidth remains at aborted bandwidth');
+  // Media may be changed, but it should be changed to the same media. In the future, this
+  // can safely not be changed.
+  assert.equal(mediaChanges.length, 3, 'changed media');
+  assert.equal(mediaChanges[2].uri, 'media.m3u8', 'media remains unchanged');
+  // A bandwidthupdate event can still be triggered, but the bandwidth hasn't changed, so
+  // the media should remain unchanged. In the future, this can safely not be triggered.
+  assert.equal(bandwidthupdateEvents, 3, 'triggered a bandwidthupdate');
+
+  segmentRequest = this.requests[0];
+  assert.equal(segmentRequest.uri.substring(segmentRequest.uri.length - 4),
```

Comment:
I think it would be good to finish this request as uncached and observe that bandwidth values are updated and a media change happens